Split clean-up into separate job

WalterKolczynski-NOAA opened the pull request:
# Description
Moves the clean-up that was previously done in the archive jobs into their own separate job. The clean-up is also streamlined considerably by using only `COM_TOP` instead of going through every template. There is also additional streamlining/corrections in the function that does the actual removing.

Some settings used by both jobs were elevated to `config.base`. Others only needed for cleanup were moved to the new config for that job.

Also corrects a small error encountered when attempting to rerun an ensemble forecast.

Resolves #583
Resolves #1872 

# Type of change
- Bugfix
- Maintenance (code refactor, clean-up, new CI test, etc.)

# Change characteristics
- Is this a breaking change (a change in existing functionality)? NO
- Does this change require a documentation update? NO

# How has this been tested?
- Cycled test on Orion

# Checklist
- [x] Any dependent changes have been merged and published
- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] My changes generate no new warnings
- [ ] New and existing tests pass with my changes
- [x] I have made corresponding changes to the documentation if necessary

WalterKolczynski-NOAA committed:
Split clean-up into separate job

Moves the clean-up that was previously done in the archive jobs into
their own separate job. It is mostly a copy of what existed in the
archive scripts, except some adjustments to account for ensemble
members in the same job as the deterministic and some others for
shellcheck compliance/other standard code improvements.

Some settings used by both jobs were elevated to `config.base`.
Others only needed for cleanup were moved to the new config for that
job.

Resolves #583

WalterKolczynski-NOAA commented:
Still working on testing and cleaning things up, but wanted to get a draft up for Rahul to look at if he has time before he goes on leave.

WalterKolczynski-NOAA committed:
Mute a couple shellcheck warnings in cleanup

aerorahul commented on jobs/JGLOBAL_CLEANUP:
```diff
@@ -0,0 +1,17 @@
+#! /usr/bin/env bash
+
+source "${HOMEgfs}/ush/preamble.sh"
+source "${HOMEgfs}/ush/jjob_header.sh" -e "cleanup" -c "base"
```
Do you not need to source `config.cleanup` after `config.base`?

WalterKolczynski-NOAA replied:
I should have to, yes. Will fix.

WalterKolczynski-NOAA replied:
Done

WalterKolczynski-NOAA replied:
Backed this change out. I think this is what caused the failure because setpy.sh isn't available.

aerorahul commented on scripts/exglobal_cleanup.sh:
```diff
@@ -0,0 +1,214 @@
+GDATE=$(${NDATE} -"${assim_freq}" "${PDY}${cyc}")
+
+# PREVIOUS to the PRIOR CYCLE
+GDATE=$(${NDATE} -"${assim_freq}" "${GDATE}")
```
Please use `date --utc ...`

WalterKolczynski-NOAA replied:
Done

aerorahul commented on scripts/exglobal_cleanup.sh:
```diff
@@ -0,0 +1,214 @@
+
+if [[ "${RUN}" == "gfs" ]]; then
+    fhmax=$((FHMAX_FITS + 36))
+    RDATE=$(${NDATE} -"${fhmax}" "${PDY}${cyc}")
```
please use `date --utc ...`

aerorahul commented on jobs/rocoto/cleanup.sh:
```diff
@@ -0,0 +1,19 @@
+
+###############################################################
+# Source FV3GFS workflow modules
+. "${HOMEgfs}"/ush/load_fv3gfs_modules.sh
+status=$?
+[[ ${status} -ne 0 ]] && exit "${status}"
```
What modules are needed for `cleanup`?

WalterKolczynski-NOAA replied:
Removed

aerorahul commented on scripts/exglobal_cleanup.sh:
```diff
@@ -0,0 +1,214 @@
+# Retain files needed by Fit2Obs
+# TODO: This whole section needs to be revamped to remove marine component
+#  directories and not look at the rocoto log.
+GDATEEND=$(${NDATE} -"${RMOLDEND:-24}"  "${PDY}${cyc}")
+GDATE=$(${NDATE} -"${RMOLDSTD:-120}" "${PDY}${cyc}")
+RTOFS_DATE=$(${NDATE} -48 "${PDY}${cyc}")
```
please use `date --utc ...`

aerorahul commented on scripts/exglobal_cleanup.sh:
```diff
@@ -0,0 +1,214 @@
+        find "${target_dir}" -empty -type d -delete
+    fi
+
+    GDATE=$(${NDATE} +"${assim_freq}" "${GDATE}")
```
please use `date --utc ...`

aerorahul commented on scripts/exglobal_cleanup.sh:
```diff
@@ -0,0 +1,214 @@
+fi
+
+# Remove $RUN.$rPDY for the older of GDATE or RDATE
+GDATE=$(${NDATE} -"${RMOLDSTD:-120}" "${PDY}${cyc}")
+RDATE=$(${NDATE} -"${FHMAX_GFS}" "${PDY}${cyc}")
```
please use `date --utc ...`

aerorahul reviewed (changes requested):
Looks ok to me.  
1. please use `date --utc ...` instead of `$NDATE`

I assume `earc.sh` will have its own cleanup; e.g. `ecleanup.sh`

HenryRWinterbottom commented on scripts/exgdas_enkf_earc.sh:
```diff
@@ -133,172 +133,4 @@ if [ "${ENSGRP}" -eq 0 ]; then
-# ENSGRP 0 also does clean-up
-###############################################################
-if [[ "${ENSGRP}" -eq 0 ]]; then
-    function remove_files() {
-        # TODO: move this to a new location
-        local directory=$1
-        shift
-        if [[ ! -d ${directory} ]]; then
-            echo "No directory ${directory} to remove files from, skiping"
-            return
-        fi
-        local exclude_list=""
-        if (($# > 0)); then
-            exclude_list=$*
-        fi
-        local file_list
-        declare -a file_list
-        # Suppress warnings about chained commands suppressing exit codes
-        # shellcheck disable=SC2312
-        readarray -t file_list < <(find -L "${directory}" -type f)
-        if (( ${#file_list[@]} == 0 )); then return; fi
-        for exclude in ${exclude_list}; do
-            echo "Excluding ${exclude}"
-            declare -a file_list_old=("${file_list[@]}")
-            # Suppress warnings about chained commands suppressing exit codes
-            # shellcheck disable=SC2312
-            readarray file_list < <(printf -- '%s\n' "${file_list_old[@]}" | grep -v "${exclude}")
-            if (( ${#file_list[@]} == 0 )); then return; fi
-        done
-
-        for file in "${file_list[@]}"; do
-            rm -f "${file}"
-        done
-        # Remove directory if empty
-        rmdir "${directory}" || true
-    }
```
@WalterKolczynski-NOAA I am addressing this function while working with Catherine Thomas on issue #1872. I have asked her to test a more robust solution to `remove_files` (see [here](https://github.com/HenryWinterbottom-NOAA/global-workflow/blob/bug/gdas_enkf_scrubbing/scripts/exgdas_enkf_earc.sh#L181)).

Would you prefer that I freeze her issue for now until we determine which is the optimal path forward?

WalterKolczynski-NOAA replied:
No, go ahead and keep working on it. The function is basically unchanged, so any solutions you find can then be applied here.

KateFriedman-NOAA commented on parm/config/gfs/config.base.emc.dyn:
```diff
@@ -395,6 +395,9 @@ export ARCH_CYC=00           # Archive data at this cycle for warm_start capabil
 export ARCH_FCSTICFREQ=1     # Archive frequency in days for gdas and gfs forecast-only capability
 
-export DELETE_COM_IN_ARCHIVE_JOB="YES"   # NO=retain ROTDIR.  YES default in arch.sh and earc.sh.
+#--online archive of nemsio files for fit2obs verification
+export FITSARC="YES"
+export FHMAX_FITS=132
+[[ "${FHMAX_FITS}" -gt "${FHMAX_GFS}" ]] && export FHMAX_FITS=${FHMAX_GFS}
```
Why move these out of `config.arch`?

WalterKolczynski-NOAA replied:
`FHMAX_FITS` is being used by cleanup as well. `FITSARC` looks like it can move back though.

KateFriedman-NOAA replied:
Wondering also if all of these variables should move to `config.fit2obs`? Would need the cleanup job to source that config as well though I think, for `FHMAX_FITS`.

KateFriedman-NOAA commented on parm/config/gfs/config.cleanup:
```diff
@@ -0,0 +1,15 @@
+# Get task specific resources
+source "${EXPDIR}/config.resources" cleanup
+
+export DELETE_COM_IN_ARCHIVE_JOB="YES"   # NO=retain ROTDIR.  YES default in cleanup.sh
```
Should this be `DELETE_COM="YES"` and be moved to `config.base`?

WalterKolczynski-NOAA replied:
Renamed, but kept it here.

KateFriedman-NOAA commented on scripts/exglobal_cleanup.sh:
```diff
@@ -0,0 +1,214 @@
+COMIN="${DATAROOT}/${GDATE}"
+[[ -d ${COMIN} ]] && rm -rf "${COMIN}"
+
+if [[ "${DELETE_COM_IN_ARCHIVE_JOB:-YES}" == NO ]] ; then
+    exit 0
+fi
```
See other comment about whether this variable should be something like `DELETE_COM`.

I'm guessing this check is here because the cleanup job will always be in the mesh but users can turn off cleanup for whatever reason?

WalterKolczynski-NOAA replied:
Yeah. Since cleanup does very little with it off, in the future we probably should just turn off the job. But I'm leaving that outside the scope of this PR.

KateFriedman-NOAA commented on scripts/exglobal_cleanup.sh:
```diff
@@ -0,0 +1,214 @@
+#  directories and not look at the rocoto log.
+GDATEEND=$(${NDATE} -"${RMOLDEND:-24}"  "${PDY}${cyc}")
+GDATE=$(${NDATE} -"${RMOLDSTD:-120}" "${PDY}${cyc}")
+RTOFS_DATE=$(${NDATE} -48 "${PDY}${cyc}")
```
Perhaps make `RTOFS_DATE` look-back configurable as well? `RMOLDRTOFS` or something similar and make 48hrs the default?

WalterKolczynski-NOAA replied:
Added

KateFriedman-NOAA reviewed (commented):
Looks great, thanks @WalterKolczynski-NOAA ! Left a few comments for now. Happy to re-review and approve when this is out of draft form.

WalterKolczynski-NOAA commented:
> I assume `earc.sh` will have its own cleanup; e.g. `ecleanup.sh`

I merged cleanup of the enkf directories into the regular cleanup, so there is no need to run a separate one for enkf.

CatherineThomas-NOAA commented:
@WalterKolczynski-NOAA 

> I merged cleanup of the enkf directories into the regular cleanup, so there is no need to run a separate one for enkf.

Should the new cleanup job also have a dependency on the completion of earc?  Looks like just arch currently.

WalterKolczynski-NOAA commented:
> @WalterKolczynski-NOAA
> 
> > I merged cleanup of the enkf directories into the regular cleanup, so there is no need to run a separate one for enkf.
> 
> Should the new cleanup job also have a dependency on the completion of earc? Looks like just arch currently.

Yes! Good catch.

HenryWinterbottom-NOAA commented:
@CatherineThomas-NOAA I have been unable to reproduce your result so I am not sure if something in the runtime environment is different or what.

However, I did build a regression test for the `remove_files` function and tried to catch any edge-cases (perhaps yours is one).

Can you please update your branch and try again? If this does not solve the problem we may have to see what PR #1906 yields.

Thank you for your patience.

WalterKolczynski-NOAA committed:
Have cleanup job source cleanup config

WalterKolczynski-NOAA committed:
Replace NDATE with posix date command in cleanup job

Refs #583

WalterKolczynski-NOAA committed:
Rename control variable for cleaning up COM

Since clean-up is no longer part of the archive job, the name of
the variable that controls whether COM is cleaned up is renamed.

Since the clean-up job does very little if this is NO, in the
future may want to just not run this job in that situation.

Refs #583

WalterKolczynski-NOAA committed:
Fix syntax of date calls in cleanup

WalterKolczynski-NOAA committed:
Add ensemble archive as prerequisite for gdas cleanup

WalterKolczynski-NOAA committed:
Streamline cleanup script

The old cleanup script was more convoluted than it needed to be.
The script has been streamlined quite a bit and now works properly.
Instead of going through every COM template, `COM_TOP is used along
with find to get everything at once (well, a few passes of find are
needed to deal with regular files, symlinks, and then directories).

Refs #583
Resolves #1872

WalterKolczynski-NOAA committed:
Fix error in rerunning a failed ensemble member

When trying to rerun a failed ensemble member, the script would fail
due to `set -u` when grep returned non-zero for not being able to
find `PASS` in the group status file. This is resolved by turning it
off and then back on.

WalterKolczynski-NOAA committed:
Add enkf back into cleanup

The streamlining of the cleanup script resulted in enkf not being
cleaned up. Script not runs for `$RUN` and `enkf${RUN}`. Since the
existence of the directory is checked before trying to clean up,
it doesn't matter if `enkf${RUN}` is not a valid `$RUN` for the
experiment.

Refs #583

WalterKolczynski-NOAA committed:
Add variable to control last RTOFS kept

WalterKolczynski-NOAA commented:
Completely revamped the cleanup to be less convoluted and actually work properly. This should resolve #1872 now as well.

WalterKolczynski-NOAA commented:
Rerunning some testing after I ran into an issue with `find`, but I think it is close enough for review now.

WalterKolczynski-NOAA committed:
Add wildcards to cleanup excludes

The `-name` option to find works differently than `grep`, so wild-
cards need to be added to make sure files to exclude are matched
properly.

HenryRWinterbottom reviewed (commented):
@WalterKolczynski-NOAA This LGTM. If it is confirmed that this also resolves #1872, I approve.

@CatherineThomas-NOAA Can you confirm whether this resolves issue #1872?

CatherineThomas-NOAA commented:
@HenryWinterbottom-NOAA 
I started a test with Walter's fork this morning, but it needs to run a few more cycles.  I'll let you know once it's gone far enough.

HenryRWinterbottom commented:
Great! Thank you @CatherineThomas-NOAA.

WalterKolczynski-NOAA commented:
After sleeping on it, I think now that the clean-up is streamlined, it will be better to have enkf run its own cleanup after all. Will make the change today. Shouldn't invalidate any running tests.

WalterKolczynski-NOAA committed:
Give enkf its own cleanup job

Split cleanup for enkf off into it's own rocoto job to make the
script cleaner. This eliminates the need to loop over RUNs and
avoids the unnecessary check of non-existent RUNs.

Refs #583

CatherineThomas-NOAA commented:
I ran two days worth of cycles with the single cleanup job (before Walter's split at [bde3297](https://github.com/NOAA-EMC/global-workflow/pull/1906/commits/bde329759a14db5e4d6dd03484e54b04a7a4d3a6)).  The files were cleaned up mostly as expected, fixing the problem with issue #1872.

The only hiccup was in the exclude file list.  It looks like the exclude file list is being shared between the deterministic and ensemble:
```
+ exglobal_cleanup.sh[47]: find /scratch1/NCEPDEV/stmp4/Catherine.Thomas/ROTDIRS/cleanup/enkfgdas.20211221/00 -type l -not '(' -name '*prepbufr*' -or -name '*prepbufr*' -or -name '*cnvstat*' -or -name '*prepbufr*' -or -name '*prepbufr*' -or -name '*cnvstat*' -or -name '*atmanl.nc' ')' -delete

```
The only file left in my ensemble directory is the cnvstat ensmean file:
```
/scratch1/NCEPDEV/stmp4/Catherine.Thomas/ROTDIRS/cleanup/enkfgdas.20211221/06/ensstat/analysis/atmos
[Catherine.Thomas@hfe01 atmos]$ ls
enkfgdas.t06z.cnvstat.ensmean
```

I don't think this is the behavior that we want.  We should probably have the option of having separate exclude lists for deterministic and ensemble.

Being that I tested this fork before @WalterKolczynski-NOAA split the cleanup job into det and ens, I'll pull his new changes and try one more cycle, but it doesn't look to me that it will make a difference for this particular issue.

CatherineThomas-NOAA commented:
@HenryWinterbottom-NOAA @WalterKolczynski-NOAA 
I ran an additional cycle with [bde3297](https://github.com/NOAA-EMC/global-workflow/pull/1906/commits/bde329759a14db5e4d6dd03484e54b04a7a4d3a6) and it performed as expected -- it ran separate gdas and enkfgdas cleanup jobs that cleaned up all the files except for the exclude list files.  As anticipated, this change did not address the single exclude list for both the deterministic and ensemble.

WalterKolczynski-NOAA commented:
@CatherineThomas-NOAA I went back and checked what it was before I changed it in the COM reorg. It looks like the exclude for ENKF was `f006.ens`. If you confirm that is correct I will make the change.

HenryRWinterbottom commented:
@WalterKolczynski-NOAA That was the only one that I saw as well.

CatherineThomas-NOAA commented:
@WalterKolczynski-NOAA 
Yes, that's right. To save the F006 ensmean and enspread.

WalterKolczynski-NOAA committed:
Use different exclude lists for cleanup

Changes the exclude list into a configurable variable as a comma-
separated string. This allows for the use of different strings for
ensemble and deterministic.

Refs #583
Refs #1872

WalterKolczynski-NOAA commented:
@CatherineThomas-NOAA Updated to set the list in `config.cleanup` and base it on `$RUN`. Note if you test using an existing experiment, you will have to similarly update the `config.cleanup` to match.

CatherineThomas-NOAA reviewed (approved):
I merged @WalterKolczynski-NOAA's changes and ran another cycle.  The different exclude lists for gdas and enkfgdas work as intended.  It even cleaned up the older cycle’s enkf cnvstats that were erroneously saved.  Looks good to me.  

Thanks @HenryWinterbottom-NOAA and @WalterKolczynski-NOAA for your help on this.

HenryRWinterbottom commented:
@CatherineThomas-NOAA Thanks for following up and also for testing.

HenryRWinterbottom reviewed (approved):
Thank you @WalterKolczynski-NOAA for resolving two issues with this PR.

TerrenceMcGuinness-NOAA commented:
This PR needs to be updated with recent changes in the CI scripts in the development branch.  This is causing the creation of the EXPDIRs to fail in the automated CI scripts.

KateFriedman-NOAA commented on parm/config/gfs/config.cleanup:
```diff
@@ -0,0 +1,25 @@
+
+#--starting and ending hours of previous cycles to be removed from rotating directory
+export RMOLDSTD=144
+export RMOLDEND=24
```
Add `RMOLDRTOFS=48` here?

KateFriedman-NOAA reviewed (approved):
One comment left. Looks good otherwise, thanks @WalterKolczynski-NOAA !

WalterKolczynski-NOAA commented:
@aerorahul Can you approve so I can merge without overriding restrictions?

aerorahul commented on workflow/rocoto/gfs_tasks.py:
```diff
@@ -944,6 +944,23 @@ def arch(self):
+    def cleanup(self):
+        deps = []
+        if 'enkf' in self.cdump:
+            dep_dict = {'type': 'metatask', 'name': 'enkfgdaseamn'}
```
Do we need to add a dependency on `enkfgfseamn` when the early cycle EnKF is also ON?

aerorahul reviewed (approved):
LGTM, though I have a clarifying question regarding job dependencies of cleanup when early cycle EnKF is enabled.
